frontend/src/metabase/redux/settings.ts

```diff
@@ -35,13 +35,18 @@ export const settings = createReducer(
   },
 );
 
+type RequiredUserSettings = {
+  [K in keyof UserSettings]-?: UserSettings[K];
+};
+
+type UserSettingsMap = {
+  [K in keyof RequiredUserSettings]: { key: K; value: RequiredUserSettings[K] };
+}[keyof RequiredUserSettings];
+
 export const UPDATE_USER_SETTING = "metabase/settings/UPDATE_USER_SETTING";
 export const updateUserSetting = createThunkAction(
   UPDATE_USER_SETTING,
-  function <K extends keyof UserSettings>(setting: {
-    key: K;
-    value: UserSettings[K];
-  }) {
+  function (setting: UserSettingsMap) {
```

Comment:
We can do it without `RequiredUserSettings` and `UserSettingsMap`.
I think this is simpler:

```suggestion
  function <K extends keyof UserSettings>(setting: {
    key: K;
    value: Exclude<UserSettings[K], undefined>;
  }) {
```

Reply:
Thanks for the thorough review, as always @kamilmielnik !
Really appreciate it.

I've applied your suggested changes and the diff is indeed much smaller. 🎉 

As a side note:
I personally really like and even prefer the explicit intellisense preview that the previous solution provided. But this a very biased and a very personal preference.